### ISSUE TITLE
Errors shown but not resolved

### ISSUE THREAD

**Issue by drjonz** [ikajaste/Joomla-2.5-Asset-Weaver#9]
I get this when I run the script:

Errors: 3 (3)
{2} LFT value not what expected by parent (exp 2 is 1)
{10821} LFT value not what expected by parent (exp 2462 is 2464)
{16085} LFT value not what expected by parent (exp 36775 is 37128)

Any idea what this means?  For another of my DBs I got this:

Errors: 2 (2)
{2} LFT value not what expected by parent (exp 2 is 1)
{4349} LFT value not what expected by parent (exp 8893 is 9246)

Adding OK doesn't fix the errors.

**Comment by ikajaste** [comment]
Short: I know what it means, but no idea if it affects Joomla in any way. :) The script _should_ fix them, seems like a bug.

Long:

The reported problem is with the tree structure. Each asset node has _left_ and _right_ values that are probably used to traverse the tree srtucture in some what. _Left_ and _right_ sort of determine the boundary that node _and it's children_ fit inside. _Left_ is a number that is one lower than the _left_ of the first child, and _right_ is one higher than the _right_ of the last child. I found a nice [example image](http://www.artfulsoftware.com/mysqlbook/sampler/20fig03.png) from [this link](http://www.artfulsoftware.com/mysqlbook/sampler/mysqled1ch20.html)

The error output first identifies the node in question `{id}` then reports the numbers that fail the check in the parentheses. `exp 2462` would mean the parent of the node has a _left_ value of 2461, so the child should have +1 to that, 2462. `is 2464` determines what the value actually was.

However, I don't know if Joomla expects theses values to be correct or not. Could be it doesn't care if they're a bit off, as long as there aren't any conflicting numbers. Though in a clean example case they are correct, so I made the script report any errors in a "better safe than sorry" -manner.

The script _should_ be able to reassign the _left_ and _right_ values for all the nodes, the fuction is called `fullReIndex()` in `AssetContainer` class... if using `ok` does nothing, that does sound like a bug in the script. I'll look into it. Probably the classe's `$reindexrequired` isn't turned to `true` in some place it should be.

**Comment by drjonz** [comment]
Thank you for the reply.

Does it affect Joomla?  I think so.  My hosting company forced me off a shared server onto VPS and is complaining again that my tiny sites with little for traffic is killing the server.  The sites don't load very fast either.  Not like they did in Joomla 1.5x.

I'll look forward to hearing if you change the script.  Or If you want me to try to edit my own, let me know.

**Comment by ikajaste** [comment]
The latest change 8cf4022bbc5b7cfb68108cb697d304d3a8e5de73 should now fix the issue by triggering the reindexing with the errors you mentioned. Hopefully the reindexing is enough to fix the asset table in your case, please report whether it worked.

**Comment by drjonz** [comment]
Reading configuration from ../configuration.php
Connecting to freedomxx as freedom_user to read tables: jos_assets
Attempting to fix article and category assets based on article category structure
Linking articles and categories done.

Errors: 3 (3)
{2} LFT value not what expected by parent (exp 2 is 1)
{4349} LFT value not what expected by parent (exp 8893 is 9246)
{1} RGT value not what expected (exp 9248 is 9250)

Applied fixes: 0
SQL to execute: 0

Reconnecting to freedom_xx as freedom_user to write changes to: jos_assets
Executing SQL:  done.

So it looks like it still doesn't attempt a fix.

**Comment by ikajaste** [comment]
Yeah, it. should say "reindexing..." and probably have huge amounts of sql to execute. Weird. I'll try to replicate the situation and fix the issue, thanks for the report.

**Comment by ikajaste** [comment]
I replicated the problem succesfully - turns out reindex wasn't always being executed when needed. Latest update f7b6da8ffdc2ee99209d9525cd47c59dfa94ca0d fixes the issue, and it now fixed the table in my replicated situation correctly. Please report whether this version worked for you.

**Comment by drjonz** [comment]
Looks the same.  This was using fresh full download of both files:

Reading configuration from ../configuration.php
Connecting to freedom_xx as freedom_xxu to read tables: jos_assets
Attempting to fix article and category assets based on article category structure
Linking articles and categories done.

Errors: 3 (3)
{2} LFT value not what expected by parent (exp 2 is 1)
{4349} LFT value not what expected by parent (exp 8897 is 9250)
{1} RGT value not what expected (exp 9252 is 9254)

Applied fixes: 0
SQL to execute: 0

Reconnecting to freedom_xx as freedom_xxu to write changes to: jos_assets
Executing SQL:  done.

**Comment by drjonz** [comment]
Wait, scratch that.  Didn't save new files over the old.  Stand by.

**Comment by drjonz** [comment]
Outstanding.  This just fixed the errors on 2 of my sites!  Many thanks.  Where's the tip jar?

**Comment by drjonz** [comment]
Tried on a 3rd site.  Oddball error.  Didn't fix it.

Reading configuration from ../configuration.php
Connecting to freedom_msn as freedom_msnu to read tables: j25_assets
Attempting to fix article and category assets based on article category structure
Linking articles and categories done.

Errors: 5 (1) Affected nodes: 4
Asset (0) not found for category (2)

Applied fixes: 0
SQL to execute: 4

**\* SKIPPING SQL EXECUTION

Reading configuration from ../configuration.php
Connecting to freedom_msn as freedom_msnu to read tables: j25_assets
Attempting to fix article and category assets based on article category structure
Linking articles and categories done.

Errors: 5 (1) Affected nodes: 4
Asset (0) not found for category (2)

Applied fixes: 0
SQL to execute: 4

Reconnecting to freedom_msn as freedom_msnu to write changes to: j25_assets
Executing SQL: .... done.

**Comment by drjonz** [comment]
Other than that odd error above, I've now fixed 6 sites.  The latest was a site upgraded with the most recent version of jupgrade out there: 2.5.1.  Wasn't expecting to see an error on that one.

**Comment by drjonz** [comment]
Here is another oddball it could not fix.

Reading configuration from ../configuration.php
Connecting to abetter_kbells as abetter_kbellsu to read tables: j25_assets
Attempting to fix article and category assets based on article category structure
Linking articles and categories done.

Errors: 16 (2) Affected nodes: 15
Asset (0) not found for category (2)
{51} Asset for category parent (0) not found

Applied fixes: 0
SQL to execute: 14
